File: engine/src/main/java/top/skyeyefast/mchjong/engine/WallLayout.java
package top.skyeyefast.mchjong.engine;

/** Public wall positions only: drawing moves clockwise, opposite to the seat order. */
public final class WallLayout {
    private WallLayout() {}

    /** Count seats from the dealer, then skip diceSum stacks from that owner's right. */
    public static int breakOffset(int dealer, int diceSum, int size, int players) {
        int stacksPerSide = size / (2 * players);
        int side = (dealer + diceSum - 1) % players;
        return 2 * (side * stacksPerSide + stacksPerSide - diceSum - 1);
    }

    public static int stack(int index, int breakOffset, int size) {
        return Math.floorMod(breakOffset / 2 - index / 2, size / 2);
    }

    public static int side(int index, int breakOffset, int size, int players) {
        return stack(index, breakOffset, size) / (size / (2 * players));
    }
}

File: engine/src/main/kotlin/top/skyeyefast/mchjong/engine/WallLayout.kt
package top.skyeyefast.mchjong.engine

/** Public wall positions only: drawing moves clockwise, opposite to the seat order. */
object WallLayout {
    /** Count seats from the dealer, then skip diceSum stacks from that owner's right. */
    @JvmStatic
    fun breakOffset(dealer: Int, diceSum: Int, size: Int, players: Int): Int {
        val stacksPerSide = size / (2 * players)
        val side = (dealer + diceSum - 1) % players
        return 2 * (side * stacksPerSide + stacksPerSide - diceSum - 1)
    }

    @JvmStatic
    fun stack(index: Int, breakOffset: Int, size: Int): Int =
        Math.floorMod(breakOffset / 2 - index / 2, size / 2)

    @JvmStatic
    fun side(index: Int, breakOffset: Int, size: Int, players: Int): Int =
        stack(index, breakOffset, size) / (size / (2 * players))
}
